### PR TITLE
Enrich assertHashContractInstance with packageName

### PR DESCRIPTION
**Pull request by tudor-da**

### CODE REVIEW

**Review by remyhaemmerle-da — APPROVED** [review]

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/crypto/Hash.scala [hunk 1]
```diff
@@ -313,21 +313,28 @@ object Hash {
...
   // The hash is not stable under suffixing of contract IDs
   @throws[HashingError]
-  def assertHashContractInstance(templateId: Ref.Identifier, arg: Value): Hash =
+  def assertHashContractInstance(
+      templateId: Ref.Identifier,
+      arg: Value,
+      packageName: Ref.PackageName = Ref.PackageName.assertFromString("default"),
+  ): Hash =
     builder(Purpose.ContractInstance, aCid2Bytes)
+      .add(packageName)
```
Comment (meiersi-da) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/crypto/Hash.scala:327]: Why does this need to be included given that, AFAIK, the template-id contains the package-hash, which in turn also hashes the package-name?

Comment (remyhaemmerle-da) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/crypto/Hash.scala:327]: One requirement of the upgrade story is that some parties should be able to use (or validate transaction included) contracts that have been created with package those parties do not  know about.  In such a case this participant rely on the fact the signatories --who necessarily know about this package-- validate that the contract was created with this particular package ***and*** that the meta-data did not change during down/upgrading. Putting the meta-data in the authenticated contract ID scheme is a convenient way for the different participants to agree on it.

Comment (meiersi-da) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/crypto/Hash.scala:327]: Ah that's a great explanation as to why we want to include the data here as well.

I suggest adding a condensed form of that explanation as a comment to the code.

**Pull request merged**